src/services/offer.js

```diff
@@ -192,29 +196,72 @@ class OfferService {
 
     /**
      * Fetches offers according to specified options
+     * Learn more about keyset search here: https://github.com/NIAEFEUP/nijobs-be/issues/129
+     *
      * @param {*} options
      * value: Text to use in full-text-search
-     * offset: Point to start looking (and limiting)
+     * queryToken: Id of the last fetched offer
      * limit: How many offers to show
      * jobType: Array of jobTypes allowed
      */
-    get({ value = "", offset = 0, limit = OfferService.MAX_OFFERS_PER_QUERY, showHidden = false, showAdminReason = false, ...filters }) {
-
-        const offers = (value ? Offer.find(
-            { "$and": [this._buildFilterQuery(filters), { "$text": { "$search": value } }] }, { score: { "$meta": "textScore" } }
-        ) : Offer.find(this._buildFilterQuery(filters))).current();
+    async get({ value = "", queryToken = null, limit = OfferService.MAX_OFFERS_PER_QUERY,
+        showHidden = false, showAdminReason = false, ...filters }) {
+
+        const {
+            id: lastOfferId,
+            score: lastOfferScore
+        } = queryToken ? this.decodeQueryToken(queryToken) : {};
+
+        let offers;
+        if (lastOfferId && value) {
+            offers = Offer.aggregate([
+                { $match: { $text: { $search: value } } },
+                { $addFields: {
+                    score: { $meta: "textScore" },
+                    adminReason: { $cond: [showAdminReason, "$adminReason", "$$REMOVE"] }
+                } },
+                { $match: { "$or": [
+                    { score: { "$lt": lastOfferScore } },
+                    { score: lastOfferScore, _id: { "$gt": ObjectId(lastOfferId) } }
+                ] } },
+                { $match: Offer.filterCurrent() },
+                { $match: showHidden ? {} : Offer.filterNonHidden() }
+            ]);
+
+        } else {
+            if (lastOfferId) {
+
+                offers = Offer.find({ "$and": [
+                    this._buildFilterQuery(filters),
+                    { _id: { "$gt": ObjectId(lastOfferId) } }
+                ] });
+
+            } else {
+
+                offers = (value ? Offer.find({ "$and": [
+                    this._buildFilterQuery(filters),
+                    { "$text": { "$search": value } }
+                ] }, { score: { "$meta": "textScore" } }
+
+                ) : Offer.find(this._buildFilterQuery(filters)));
+            }
 
-        if (!showHidden) offers.withoutHidden();
+            offers.current();
+            if (!showHidden) offers.withoutHidden();
+            if (!showAdminReason) offers.select("-adminReason");
+        }
 
-        const offersQuery = offers
-            .sort(value ? { score: { "$meta": "textScore" } } : undefined)
-            .skip(offset)
+        const results = await offers
+            .sort(value ? { score: { "$meta": "textScore" }, _id: 1 } : { _id: 1 })
```

Comment:
Is there any way can we support a different `sort by` value? I'm thinking we might want to sort results other than by `relevance`

Reply:
I think we can sort by whatever field we want, we just need an extra prop in the query and then we can choose different properties for the sort

I can create an issue for that, as it seems like a different matter

Reply:
But we'd have to send the sort by in the queryToken as well right? Otherwise we wouldn't know when loading more, or it could possibly be different

Reply:
I was thinking we could let FE choose the prop it wants to sort for (query) and then use it to encode/decode the queryToken

Reply:
That's weird, what does it mean "use it to encode/decode"? If the backend doesn't know what it is, hiw will it decode it?

Also, changing the sortby creates a new search, so we cannot search, change the sortby and send the first queryToken, it's like using different query filters.

So the sort by would neet to be part of the queryToken, and be used by the backend and the FE doesn't know anything about it, except that it will send it on first search requests, and every time it changes, issue a new (different) search request

Reply:
Well yes, changing the sortby would need a new search. I was saying something like this:

FE: sortBy: `field`
BE:
- Sorts the results by `field` and the last element has `field: X`
- Constructs the queryToken as `{id: 123, sortField: X}`

And then the flow would be the same as currently. But the FE would need a new search to change the sortBy

Reply:
Exactly, it's only a matter of making the sortby generic. Currently it's score, but it could be other fields

Reply:
Agreed. But I think we should release with infinite scrolling as it is and implement this in a separate PR

Reply:
Sure, just wanted to make sure this could be easily extended, and it seems like it is.

Please create the issue for visibility